### PR TITLE
Fix Flakiness in TestRmRestartingContainer

### LINKED ISSUE #32306: `TestRmRestartingContainer` integration test is flaky.
I think the `TestRmRestartingContainer` introduced [here](https://github.com/docker/docker/pull/30870) has an inherent race condition. It looks for a specific string in the error message:

> cannot remove a restarting container

But the container will be continuously thrashing between an "running" and "restarting" state. I'm running into this race condition in another PR I have.

Failures look like this:

```
----------------------------------------------------------------------
FAIL: docker_cli_rm_test.go:91: DockerSuite.TestRmRestartingContainer

docker_cli_rm_test.go:97:
    c.Assert(res, checker.Contains, "cannot remove a restarting container")
... obtained string = "Error response from daemon: You cannot remove a running container deadfbb730efd22997cfc43de1317163adee97afa2e47021420c622cde96816e. container before attempting removal or force remove\n"
... substring string = "cannot remove a restarting container"


----------------------------------------------------------------------
```

It just so happens that at the moment that the container remove request gets handled, the container happened to be running the command and not in a restarting state.

### PR DESCRIPTION
**Pull request by jlhawn**
fixes https://github.com/docker/docker/issues/32306

### CODE REVIEW

**Review by AkihiroSuda — APPROVED** [review]
LGTM

cc @vdemeester

**Review by thaJeztah — APPROVED** [review]
LGTM to make it not flaky

we can probably rewrite this to a unit test. I'll have a look at that

**Comment by thaJeztah** [comment]
opened https://github.com/docker/docker/pull/32312 to make these unit tests, but let me know if you think they would still be valuable as an integration test

**Review by vdemeester — APPROVED** [review]
LGTM 🐸

**Comment by vdemeester** [comment]
@thaJeztah in your PR, I think you could remove the integration test then 👼

File: integration-cli/docker_cli_rm_test.go [hunk 1]
```diff
@@ -93,6 +95,17 @@ func (s *DockerSuite) TestRmRestartingContainer(c *check.C) {
...
 
 	res, _, err := dockerCmdWithError("rm", name)
+
+	for strings.Contains(res, "cannot remove a running container") {
+		// The `date` command hasn't exited yet. It should end
+		// up in a "restarting" state if we wait a bit, though
+		// it is possible that it might be running again by
+		// that time. The wait period between each restart
+		// increases though so we just loop in this condition.
+		time.Sleep(100 * time.Millisecond)
```
Comment (ripcurld0) [inline integration-cli/docker_cli_rm_test.go]: I am afraid `time.Sleep` isn't good enough.
What will happen if after the 100 msecs the date will be restarted and the container will be running. The test will be still running.

**Review by boaz0 — CHANGES_REQUESTED** [review]
I think adding `trialsNo` in `for` is good.
Also, maybe it will take less time to run `exit 0` instead of `date`.
I am still not comfortable about this test though.

Sorry I didn't see this happens.

File: integration-cli/docker_cli_rm_test.go (hunk 1 above)
Comment (vdemeester) [inline integration-cli/docker_cli_rm_test.go]: @ripcurld0 as @jlhawn commented, the `wait` period increases between each restart, so if we don't catch it the first time, we'll catch it the 2nd time.. Looking at the build logs, it didn't take too much time so I think we are safe to merge 👍

**Pull request merged**